mindmeld/components/_util.py

```diff
@@ -38,3 +43,212 @@ def _get_module_or_attr(module_name: str, func_name: str = None):
     if func_name not in dir(m):
         raise ImportError(f"Cannot import {func_name} from {module_name}")
     return getattr(m, func_name)
+
+
+class TreeNode:
+    def __init__(self, nlp_name: str, parent=None, children=None, allow=None):
+        """
+        Constructor for the tree node
+        Args:
+            nlp_name: The name of the NLP component
+            parent: The parent of the NLP component. eg. parent of
+                an intent is a domain
+            children: The children of the NLP component. eg.
+                children of an intent are entities
+            allow: If True, the NLP component will be considered for inference,
+                if False, the component will not be considered for inference
+        """
+        self.nlp_name = nlp_name
+        self.allow = allow
+        self.parent = parent
+        self.children = children or []
+
+
+class TreeNlp:
+    def __init__(self, nlp, allow=None):
+        # root
+        self.root = TreeNode('root', allow=allow, children=[])
+        # construct NLP tree
+        for domain in nlp.domains:
+            domain_node = TreeNode(domain, parent=self.root, allow=allow, children=[])
+            self.root.children.append(domain_node)
+            for intent in nlp.domains[domain].intents:
+                intent_node = TreeNode(intent, parent=domain_node, allow=allow, children=[])
+                domain_node.children.append(intent_node)
+                entities = nlp.domains[domain].intents[intent].entities
+                for entity in entities:
+                    entity_node = TreeNode(entity, parent=intent_node, allow=allow, children=[])
+                    intent_node.children.append(entity_node)
+                    for role in entities[entity].role_classifier.roles:
+                        role_node = TreeNode(role, parent=intent_node, allow=allow, children=None)
+                        entity_node.children.append(role_node)
+
+    @staticmethod
+    def _convert_tree_node_to_values(*nlp_components):
+        result = [None for _ in ['domain', 'intent', 'entity', 'role']]
+        for idx, component in enumerate(nlp_components):
+            component_name = component.nlp_name if isinstance(
+                component, TreeNode) else component
+            result[idx] = component_name
+        return result
+
+    def get_domain_nodes(self):
+        return self.root.children or []
+
+    def get_intent_nodes(self, domain: Union[str, TreeNode]):
+        domain, _, _, _ = self._convert_tree_node_to_values(domain)
+        for domain_node in self.root.children:
+            if domain_node.nlp_name == domain:
+                return domain_node.children
+        return []
+
+    def get_entity_nodes(self, domain: Union[str, TreeNode],
+                         intent: Union[str, TreeNode]):
+        domain, intent, _, _ = self._convert_tree_node_to_values(domain, intent)
```

Comment:
I think this operation is performant in the base case since all of the values are strings in the `update` function, so are just looping through strings and checking their type, so it should be pretty fast. Since this function is getting called elsewhere with TreeNode types, its worth to keep it without adding the extra "I have already checked the type" arg.